ddocs/medic-sms/views/messages_by_gateway_ref/map.js

```diff
@@ -0,0 +1,23 @@
+function(doc) {
+  var _emitMsg = function(msg) {
+    if (msg && msg.gateway_ref) {
+      emit(msg.gateway_ref, msg.uuid);
+    }
+  };
+
+  // outgoing
+  var _emit = function(tasks) {
+    tasks.forEach(function(task) {
+      if (task.messages) {
+        task.messages.forEach(_emitMsg);
```

Comment:
This tries to emit `gateway_ref` property from the task message, while `task-utils` saves the property in the `task` object. 
This view ends up indexing incoming SMS messages only, so we lose the `gateway_ref` <-> doc connection.

Reply:
> This tries to emit gateway_ref property from the task message, while task-utils saves the property in the task object.

Oh man, what a mess. You're absolutely right. I think technically we should be storing the state, details, and gateway_ref on the message but we're storing them all on the task. This means we won't know if the first message fails because it could be overwritten by the second message succeeding.

In practice I _think_ we only ever have one message in this array ( reference: https://github.com/medic/medic/blob/master/shared-libs/message-utils/src/index.js#L257 ) so we get away with it. One day we should confirm this is true and remove the array altogether because it makes life way too complicated.

For now though, I think we have two choices:

1. Store the gateway_ref on the message which is correct but inconsistent with the status.
2. Update the view to get the gateway_ref from the task which is wrong but consistent.

I'm leaning towards option 2 if you're ok with that? Either way I'll update an e2e test to ensure the gateway_ref is where it needs to be.

> This view ends up indexing incoming SMS messages only, so we lose the gateway_ref <-> doc connection.

It used to only do incoming but on line 12 and 13 I added two lines to emit outgoing as well.

Reply:
I also think that there's always just one message in the list. 
The reason why messages is an array is a mystery to me, I always thought that it's a relic from earlier times. And yea, definitely we should remove the array. 
Option 2 is OK for me.

Reply:
Done.